File: scripts/analyze_intent_rotation.py
#!/usr/bin/env python3
"""Paired diagnostic: change display order/code assignment, not task semantics."""
import argparse
import collections
import json
from pathlib import Path

from evaluate_intents import load, rows, save, sha
from evaluate_intents_wide import code


def main():
    parser=argparse.ArgumentParser(description=__doc__)
    parser.add_argument('root',type=Path)
    args=parser.parse_args();root=args.root;out=root/'diagnostics'
    gold={r['id']:r for r in rows(root/'prepared/gold.jsonl') if r['dataset']=='banking77-en'}
    before={r['id']:r for r in rows(root/'runs/Qwen3-8B-Q8_0/predictions.jsonl') if r['id'] in gold}
    after_rows=rows(out/'qwen-rotation38.jsonl');after={r['id']:r for r in after_rows}
    assert len(after)==len(after_rows)==len(before)==len(gold)==200
    assert set(before)==set(after)==set(gold)
    full={r['id']:r for r in rows(root/'prepared/requests.jsonl')}
    diagnostic=rows(out/'banking77-requests.jsonl')
    assert diagnostic==[r for r in full.values() if r['id'] in gold]
    assert (out/'rotation38.exit-code').read_text().strip()=='0'
    counts=collections.Counter(); pairs=[]; lengths=collections.defaultdict(collections.Counter)
    for key in gold:
        p,q=before[key],after[key]
        assert not p.get('error') and not q.get('error')
        a,b=[r['response']['results'][0] for r in (p,q)]
        assert not a['truncated'] and not b['truncated']
        ab,bb=p['response']['backend'],q['response']['backend']
        assert ab.get('code_rotation',0)==0 and bb['code_rotation']==38
        for field in ('model_path','compute','offload_device','execution_mode','prompt_layout','prompt_profile'):
            assert ab[field]==bb[field]
        assert [s['id'] for s in a['scores']]==[s['id'] for s in b['scores']]
        assert len(a['scores'])==len(b['scores'])==77
        for i,s in enumerate(b['scores']):
            assert s['code']==code((i+77-38)%77,77)
        pick=lambda r:min(r['scores'],key=lambda s:(-s['option_probability'],s['id']))
        x,y=pick(a),pick(b)
        correct_a=x['id']==gold[key]['expected'];correct_b=y['id']==gold[key]['expected']
        counts['before_correct']+=correct_a;counts['after_correct']+=correct_b
        counts['prediction_changed']+=x['id']!=y['id']
        counts['correct_to_wrong']+=correct_a and not correct_b
        counts['wrong_to_correct']+=not correct_a and correct_b
        counts['wrong_both']+=not correct_a and not correct_b
        for label,result,chosen,correct in [('before',a,x,correct_a),('after',b,y,correct_b)]:
            if chosen['option_probability']>=.9:
                counts[label+'_confidence90_count']+=1
                counts[label+'_confidence90_wrong']+=not correct
        target_a=next(s for s in a['scores'] if s['id']==gold[key]['expected'])
        target_b=next(s for s in b['scores'] if s['id']==gold[key]['expected'])
        length_key=f"{len(target_a['token_ids'])}->{len(target_b['token_ids'])}"
        lengths[length_key]['n']+=1;lengths[length_key]['before_correct']+=correct_a;lengths[length_key]['after_correct']+=correct_b
        pairs.append(dict(id=key,expected=gold[key]['expected'],before=x['id'],after=y['id'],
            before_correct=correct_a,after_correct=correct_b,before_code=x['code'],after_code=y['code'],
            gold_code_before=target_a['code'],gold_code_after=target_b['code'],
            gold_token_lengths=[len(target_a['token_ids']),len(target_b['token_ids'])]))
    summary=dict(dataset='BANKING77 English',n=200,rotation=38,counts=dict(counts),
        gold_token_length_transitions={k:dict(v) for k,v in lengths.items()},
        requests_sha256=sha(out/'banking77-requests.jsonl'),predictions_sha256=sha(out/'qwen-rotation38.jsonl'),
        scope='Exploratory paired diagnostic; same text, candidate semantics and model. Display order and code assignment change together. Not a tuned replacement benchmark score.')
    save(out/'rotation-analysis.json',summary);save(out/'paired-predictions.json',pairs)
    base=load(out/'baseline-analysis.json')
    text=['# Intent classification diagnosis', '',
        'Qwen3-8B Q8_0, RTX 3060, frozen English/Korean 200-row test subsets. Baseline: English 111/200 (55.5%); Korean 98/200 (49.0%).', '',
        '## Findings from the original run', '',
        '- No inference errors, truncation, missing classes or missing rows were observed. The source-row/label mapping, complete candidate set and probability calculations were independently checked.',
        '- English had 89 errors. Only 14 examples had a gold label assigned a two-token code, of which 5 were correct. The remaining 186 examples had one-token gold codes and only 106 were correct. Multi-token gold codes therefore cannot account for most observed errors by themselves.',
        '- Korean had 102 errors. Twelve examples had two-token gold codes (4 correct); 188 had one-token gold codes (94 correct). These groups contain different labels and examples, so the rate difference is not a causal token-length estimate.',
        '- Candidate confidence is poorly calibrated as correctness: at confidence ≥90%, English had 58 errors among 163 predictions, and Korean had 68 among 160. These are candidate-relative probabilities, not calibrated correctness probabilities.',
        '- Examples of English confusion include `failed_transfer` → `declined_transfer` (6) and `compromised_card` → `contactless_not_working` (3). Korean confusions include `play_podcasts` → `play_music`, `email_query` → `email_sendemail` and `weather_query` → `datetime_query` (3 each). Some errors are fine-grained intent distinctions; others cross clearly different intents.', '',
        '## Paired display/code rotation control', '',
        'All original 200 English inputs and all 77 label meanings were retained. Rotation 38 changes which displayed code corresponds to each label and moves labels in the displayed list. This measures combined display-order/code-assignment sensitivity; it does not isolate tokenizer length from order or wording.', '',
        f"- Original accuracy: {counts['before_correct']}/200 ({counts['before_correct']/2:.1f}%). Rotated accuracy: {counts['after_correct']}/200 ({counts['after_correct']/2:.1f}%).",
        f"- Changed predictions: {counts['prediction_changed']}/200 ({counts['prediction_changed']/2:.1f}%). Correct → wrong: {counts['correct_to_wrong']}; wrong → correct: {counts['wrong_to_correct']}.",
        f"- Rotated predictions at confidence ≥90%: {counts['after_confidence90_wrong']} errors among {counts['after_confidence90_count']} predictions.", '',
        'This is a diagnosis on the test subset, not independent validation of a better prompt. The original benchmark scores remain unchanged. Any proposed prompt, code-bias correction or calibration should be chosen on separate development/calibration data before a new held-out evaluation.', '',
        '## Interpretation', '',
        'The current setup is zero-shot intent selection from short label-name criteria, using a general language model and answer-code likelihoods. It is not a classifier trained on BANKING77/MASSIVE examples. Label semantics, option presentation, code priors and the absence of calibration are plausible contributors. The measured control quantifies ordering/code sensitivity; it does not prove which factor causes every error. More model parameters or a larger candidate limit alone do not establish better intent classification.', '',
        'Next experiments should use a separate development split: explicit intent definitions and confusing-class examples; label/order robustness controls; then held-out calibration or task-specific training. Preserve the untouched test split for final validation.', '',
        'Evidence: `baseline-analysis.json`, `rotation-analysis.json`, `paired-predictions.json`, exact input/output JSONL, GPU log and `rotate.sh`.', '']
    (out/'REPORT.md').write_text('\n'.join(text))
    print(json.dumps(summary,indent=2))


if __name__=='__main__': main()

File: crates/l2s1-tools/src/analyze_intent_rotation.rs
use crate::{
    common::{read_jsonl, sha256, write_json},
    evaluate_intents_wide::code,
};
use anyhow::{Context, Result, ensure};
use clap::Args as ClapArgs;
use serde_json::{Value, json};
use std::{
    collections::{BTreeMap, HashMap, HashSet},
    fs,
    path::PathBuf,
};

#[derive(ClapArgs)]
pub struct Args {
    root: PathBuf,
}

fn picked(result: &Value) -> Result<&Value> {
    result["scores"]
        .as_array()
        .context("missing scores")?
        .iter()
        .min_by(|a, b| {
            b["option_probability"]
                .as_f64()
                .unwrap_or(f64::NAN)
                .total_cmp(&a["option_probability"].as_f64().unwrap_or(f64::NAN))
                .then_with(|| {
                    a["id"]
                        .as_str()
                        .unwrap_or("")
                        .cmp(b["id"].as_str().unwrap_or(""))
                })
        })
        .context("empty scores")
}

pub fn run(args: Args) -> Result<()> {
    let root = args.root;
    let out = root.join("diagnostics");
    let gold_rows = read_jsonl(&root.join("prepared/gold.jsonl"))?
        .into_iter()
        .filter(|r| r["dataset"] == "banking77-en")
        .collect::<Vec<_>>();
    let gold = gold_rows
        .iter()
        .map(|r| Ok((r["id"].as_str().context("gold ID")?.to_owned(), r)))
        .collect::<Result<BTreeMap<_, _>>>()?;
    let before = read_jsonl(&root.join("runs/Qwen3-8B-Q8_0/predictions.jsonl"))?
        .into_iter()
        .filter(|r| gold.contains_key(r["id"].as_str().unwrap_or("")))
        .map(|r| Ok((r["id"].as_str().context("prediction ID")?.to_owned(), r)))
        .collect::<Result<HashMap<_, _>>>()?;
    let after_rows = read_jsonl(&out.join("qwen-rotation38.jsonl"))?;
    let after = after_rows
        .iter()
        .map(|r| Ok((r["id"].as_str().context("rotation ID")?.to_owned(), r)))
        .collect::<Result<HashMap<_, _>>>()?;
    ensure!(
        gold.len() == 200 && before.len() == 200 && after.len() == 200 && after_rows.len() == 200,
        "paired coverage differs"
    );
    ensure!(
        gold.keys().collect::<HashSet<_>>() == before.keys().collect()
            && gold.keys().collect::<HashSet<_>>() == after.keys().collect(),
        "paired IDs differ"
    );
    let requests = read_jsonl(&root.join("prepared/requests.jsonl"))?;
    let expected = requests
        .into_iter()
        .filter(|r| gold.contains_key(r["id"].as_str().unwrap_or("")))
        .collect::<Vec<_>>();
    ensure!(
        read_jsonl(&out.join("banking77-requests.jsonl"))? == expected,
        "diagnostic requests differ"
    );
    ensure!(
        fs::read_to_string(out.join("rotation38.exit-code"))?.trim() == "0",
        "rotation run failed"
    );
    let mut counts = BTreeMap::<&str, usize>::new();
    let mut lengths = BTreeMap::<String, BTreeMap<&str, usize>>::new();
    let mut pairs = Vec::new();
    for gold in &gold_rows {
        let id = gold["id"].as_str().context("gold ID")?;
        let p = &before[id];
        let q = after[id];
        ensure!(
            p.get("error").is_none() && q.get("error").is_none(),
            "runtime error"
        );
        let a = &p["response"]["results"][0];
        let b = &q["response"]["results"][0];
        ensure!(
            a["truncated"] == false && b["truncated"] == false,
            "truncated result"
        );
        let old = &p["response"]["backend"];
        let new = &q["response"]["backend"];
        ensure!(
            old["code_rotation"].as_u64().unwrap_or(0) == 0 && new["code_rotation"] == 38,
            "rotation setting changed"
        );
        for field in [
            "model_path",
            "compute",
            "offload_device",
            "execution_mode",
            "prompt_layout",
            "prompt_profile",
        ] {
            ensure!(old[field] == new[field], "backend setting changed: {field}");
        }
        let aa = a["scores"].as_array().context("before scores")?;
        let bb = b["scores"].as_array().context("after scores")?;
        ensure!(
            aa.len() == 77
                && bb.len() == 77
                && aa.iter().map(|s| &s["id"]).eq(bb.iter().map(|s| &s["id"])),
            "candidate set changed"
        );
        for (i, score) in bb.iter().enumerate() {
            ensure!(
                score["code"] == code((i + 77 - 38) % 77, 77),
                "rotated code differs"
            );
        }
        let x = picked(a)?;
        let y = picked(b)?;
        let expected = gold["expected"].as_str().context("gold label")?;
        let correct_a = x["id"] == expected;
        let correct_b = y["id"] == expected;
        for (key, value) in [
            ("before_correct", correct_a),
            ("after_correct", correct_b),
            ("prediction_changed", x["id"] != y["id"]),
            ("correct_to_wrong", correct_a && !correct_b),
            ("wrong_to_correct", !correct_a && correct_b),
            ("wrong_both", !correct_a && !correct_b),
        ] {
            *counts.entry(key).or_default() += usize::from(value);
        }
        for (label, chosen, correct) in [("before", x, correct_a), ("after", y, correct_b)] {
            if chosen["option_probability"]
                .as_f64()
                .context("probability")?
                >= 0.9
            {
                *counts
                    .entry(if label == "before" {
                        "before_confidence90_count"
                    } else {
                        "after_confidence90_count"
                    })
                    .or_default() += 1;
                *counts
                    .entry(if label == "before" {
                        "before_confidence90_wrong"
                    } else {
                        "after_confidence90_wrong"
                    })
                    .or_default() += usize::from(!correct);
            }
        }
        let target_a = aa
            .iter()
            .find(|s| s["id"] == expected)
            .context("gold candidate absent")?;
        let target_b = bb
            .iter()
            .find(|s| s["id"] == expected)
            .context("gold candidate absent")?;
        let first = target_a["token_ids"].as_array().context("token IDs")?.len();
        let second = target_b["token_ids"].as_array().context("token IDs")?.len();
        let lengths_entry = lengths.entry(format!("{first}->{second}")).or_default();
        for (key, value) in [
            ("n", true),
            ("before_correct", correct_a),
            ("after_correct", correct_b),
        ] {
            *lengths_entry.entry(key).or_default() += usize::from(value);
        }
        pairs.push(json!({"id":id,"expected":expected,"before":x["id"],"after":y["id"],"before_correct":correct_a,"after_correct":correct_b,
            "before_code":x["code"],"after_code":y["code"],"gold_code_before":target_a["code"],"gold_code_after":target_b["code"],
            "gold_token_lengths":[first,second]}));
    }
    let summary = json!({"dataset":"BANKING77 English","n":200,"rotation":38,"counts":counts,"gold_token_length_transitions":lengths,
        "requests_sha256":sha256(&out.join("banking77-requests.jsonl"))?,"predictions_sha256":sha256(&out.join("qwen-rotation38.jsonl"))?,
        "scope":"Exploratory paired diagnostic; same text, candidate semantics and model. Display order and code assignment change together. Not a tuned replacement benchmark score."});
    write_json(&out.join("rotation-analysis.json"), &summary)?;
    write_json(&out.join("paired-predictions.json"), &json!(pairs))?;
    let c = &summary["counts"];
    let lines=vec!["# Intent classification diagnosis".to_owned(),String::new(),
        "Qwen3-8B Q8_0 on the same frozen 200 English BANKING77 requests; the diagnostic rotates displayed candidate order and answer-code assignment together.".to_owned(),String::new(),
        format!("Original correct: {}/200. Rotated correct: {}/200. Changed predictions: {}/200. Correct to wrong: {}; wrong to correct: {}.",
            c["before_correct"],c["after_correct"],c["prediction_changed"],c["correct_to_wrong"],c["wrong_to_correct"]),String::new(),
        format!("At rotated confidence >=90%: {} wrong among {} predictions.",c["after_confidence90_wrong"],c["after_confidence90_count"]),String::new(),
        "This measures combined order and code sensitivity on an existing test subset. It does not isolate tokenizer length, establish a better prompt, or replace the original benchmark score.".to_owned(),String::new(),
        "`rotation-analysis.json` and `paired-predictions.json` retain all paired outcomes and token-length transitions.".to_owned(),String::new()];
    fs::write(out.join("REPORT.md"), lines.join("\n"))?;
    println!("{}", serde_json::to_string_pretty(&summary)?);
    Ok(())
}
